File: C#/WindowsPhone/LGame-XNA-lib/Loon.Core.Graphics.Opengl/LSTRDictionary.cs
using System.Text;
using System.Collections.Generic;
using Loon.Utils;
namespace Loon.Core.Graphics.Opengl
{
    public class LSTRDictionary
    {

        private static readonly System.Collections.Generic.Dictionary<string, LFont> cacheList = new System.Collections.Generic.Dictionary<string, LFont>(
                20);

        private static readonly System.Collections.Generic.Dictionary<LFont, Dict> fontList = new System.Collections.Generic.Dictionary<LFont, Dict>(
                20);

        private static System.Collections.Generic.Dictionary<string, LSTRFont> lazyEnglish = new System.Collections.Generic.Dictionary<string, LSTRFont>(
                10);

        public static readonly string added = "0123456789";

        public const char split = '$';

        private static StringBuilder lazyKey;

        public class Dict : LRelease
        {

            internal List<char> dicts;

            internal LSTRFont font;

            static internal Dict NewDict()
            {
                return new Dict();
            }

            internal Dict()
            {
                dicts = new List<char>(512);
            }

            public void Dispose()
            {
                if (font != null)
                {
                    font.Dispose();
                    font = null;
                }
                if (dicts != null)
                {
                    dicts.Clear();
                    dicts = null;
                }
            }

        }

        public static void ClearStringLazy()
        {
            lock (cacheList)
            {
                if (cacheList != null)
                {
                    cacheList.Clear();
                }
            }
            lock (fontList)
            {
                foreach (Dict d in fontList.Values)
                {
                    if (d != null)
                    {
                        d.Dispose();
                    }
                }
                fontList.Clear();
            }
        }

        private static readonly int size = LSystem.DEFAULT_MAX_CACHE_SIZE * 5;

        public static Dict Bind(LFont font, string mes)
        {
            string message = mes + added;
            if (cacheList.Count > size)
            {
                ClearStringLazy();
            }
            lock (fontList)
            {
                LFont cFont = (LFont)CollectionUtils.Get(cacheList,message);
                Dict pDict = (Dict)CollectionUtils.Get(fontList,font);
                if (cFont == null || pDict == null)
                {
                    if (pDict == null)
                    {
                        pDict = Dict.NewDict();
                       CollectionUtils.Put(fontList,font, pDict);
                    }
                    lock (pDict)
                    {
                        CollectionUtils.Put(cacheList,message, font);
                        List<char> charas = pDict.dicts;
                        int oldSize = charas.Count;
                        char[] chars = message.ToCharArray();
                        for (int i = 0; i < chars.Length; i++)
                        {
                            if (!charas.Contains(chars[i]))
                            {
                                charas.Add(chars[i]);
                            }
                        }
                        int newSize = charas.Count;
                        if (oldSize != newSize)
                        {
                            if (pDict.font != null)
                            {
                                pDict.font.Dispose();
                                pDict.font = null;
                            }
                            StringBuilder sbr = new StringBuilder(newSize);
                            for (int i = 0; i < newSize; i++)
                            {
                                sbr.Append(charas[i]);
                            }
                            pDict.font = new LSTRFont(font, sbr.ToString());
                        }
                    }
                }
                return pDict;
            }
        }

        public static void DrawString(LFont font, string message, float x,
                float y, float angle, LColor c)
        {
            Dict pDict = Bind(font, message);
            if (pDict.font != null)
            {
                lock (pDict.font)
                {
              
                    pDict.font.DrawString(message, x, y, angle, c);
                }
            }
        }

        public static void DrawString(LFont font, string message, float x,
                float y, float sx, float sy, float ax, float ay, float angle,
                LColor c)
        {
            Dict pDict = Bind(font, message);
            if (pDict.font != null)
            {
                lock (pDict.font)
                {
                    pDict.font.DrawString(message, x, y, sx, sy, ax, ay, angle, c);
                }
            }
        }

        public static string MakeStringLazyKey(LFont font, string text)
        {
            int hashCode = 0;
            hashCode = LSystem.Unite(hashCode, font.GetSize());
            hashCode = LSystem.Unite(hashCode, font.GetStyle());
            if (lazyKey == null)
            {
                lazyKey = new StringBuilder();
                lazyKey.Append(font.GetFontName().ToLower());
                lazyKey.Append(hashCode);
                lazyKey.Append(split);
                lazyKey.Append(text);
            }
            else
            {
                lazyKey.Clear();
                lazyKey.Append(font.GetFontName().ToLower());
                lazyKey.Append(hashCode);
                lazyKey.Append(split);
                lazyKey.Append(text);
            }
            return lazyKey.ToString();
        }

        private static string MakeLazyWestKey(LFont font)
        {
            if (lazyKey == null)
            {
                lazyKey = new StringBuilder();
                lazyKey.Append(font.GetFontName().ToLower());
                lazyKey.Append(font.GetStyle());
                lazyKey.Append(font.GetSize());
            }
            else
            {
                lazyKey.Clear();
                lazyKey.Append(font.GetFontName().ToLower());
                lazyKey.Append(font.GetStyle());
                lazyKey.Append(font.GetSize());
            }
            return lazyKey.ToString();
        }

        public static void ClearEnglishLazy()
        {
            lock (lazyEnglish)
            {
                foreach (LSTRFont str in lazyEnglish.Values)
                {
                    if (str != null)
                    {
                        str.Dispose();
                    }
                }
            }
        }

        public static LSTRFont GetGLFont(LFont f)
        {
            if (lazyEnglish.Count > LSystem.DEFAULT_MAX_CACHE_SIZE)
            {
                ClearEnglishLazy();
            }
            string key = MakeLazyWestKey(f);
            LSTRFont font = (LSTRFont)CollectionUtils.Get(lazyEnglish, key);
            if (font == null)
            {
                font = new LSTRFont(f, true);
                CollectionUtils.Put(lazyEnglish, key, font);
            }
            return font;
        }

        public static void Dispose()
        {
            ClearEnglishLazy();
            ClearStringLazy();
        }
    }
}

File: Java/Android/OpenGLES/source/core/src/loon/core/graphics/opengl/LSTRDictionary.java
package loon.core.graphics.opengl;

import java.util.ArrayList;
import java.util.HashMap;

import loon.core.LRelease;
import loon.core.LSystem;
import loon.core.graphics.LColor;
import loon.core.graphics.LFont;

/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
public final class LSTRDictionary {

	private final static HashMap<String, LFont> cacheList = new HashMap<String, LFont>(
			20);

	private final static HashMap<LFont, Dict> fontList = new HashMap<LFont, Dict>(
			20);

	private static HashMap<String, LSTRFont> lazyEnglish = new HashMap<String, LSTRFont>(
			10);

	public final static String added = "0123456789";

	public final static char split = '$';

	private static StringBuffer lazyKey;

	static class Dict implements LRelease {

		ArrayList<Character> dicts;

		LSTRFont font;

		static Dict newDict() {
			return new Dict();
		}

		Dict() {
			dicts = new ArrayList<Character>(512);
		}

		@Override
		public void dispose() {
			if (font != null) {
				font.dispose();
				font = null;
			}
			if (dicts != null) {
				dicts.clear();
				dicts = null;
			}
		}

	}

	public static void clearStringLazy() {
		synchronized (cacheList) {
			if (cacheList != null) {
				cacheList.clear();
			}
		}
		synchronized (fontList) {
			for (Dict d : fontList.values()) {
				if (d != null) {
					d.dispose();
					d = null;
				}
			}
			fontList.clear();
		}
	}
	
	private final static int size = LSystem.DEFAULT_MAX_CACHE_SIZE * 5;

	public final static Dict bind(final LFont font, final String mes) {
		final String message = mes + added;
		if (cacheList.size() > size) {
			clearStringLazy();
		}
		synchronized (fontList) {
			LFont cFont = cacheList.get(message);
			Dict pDict = fontList.get(font);
			if (cFont == null || pDict == null) {
				if (pDict == null) {
					pDict = Dict.newDict();
					fontList.put(font, pDict);
				}
				synchronized (pDict) {
					cacheList.put(message, font);
					ArrayList<Character> charas = pDict.dicts;
					int oldSize = charas.size();
					char[] chars = message.toCharArray();
					for (int i = 0; i < chars.length; i++) {
						if (!charas.contains(chars[i])) {
							charas.add(chars[i]);
						}
					}
					int newSize = charas.size();
					if (oldSize != newSize) {
						if (pDict.font != null) {
							pDict.font.dispose();
							pDict.font = null;
						}
						StringBuffer sbr = new StringBuffer(newSize);
						for (int i = 0; i < newSize; i++) {
							sbr.append(charas.get(i));
						}
						pDict.font = new LSTRFont(font, sbr.toString());
					}
				}
			}
			return pDict;
		}
	}

	public final static void drawString(LFont font, String message, float x,
			float y, float angle, LColor c) {
		Dict pDict = bind(font, message);
		if (pDict.font != null) {
			synchronized (pDict.font) {
				pDict.font.drawString(message, x, y, angle, c);
			}
		}
	}

	public final static void drawString(LFont font, String message, float x,
			float y, float sx, float sy, float ax, float ay, float angle,
			LColor c) {
		Dict pDict = bind(font, message);
		if (pDict.font != null) {
			synchronized (pDict.font) {
				pDict.font.drawString(message, x, y, sx, sy, ax, ay, angle, c);
			}
		}
	}

	/**
	 * 生成特定字符串的缓存用ID
	 * 
	 * @param font
	 * @param text
	 * @return
	 */
	public static String makeStringLazyKey(final LFont font, final String text) {
		int hashCode = 0;
		hashCode = LSystem.unite(hashCode, font.getSize());
		hashCode = LSystem.unite(hashCode, font.getStyle());
		if (lazyKey == null) {
			lazyKey = new StringBuffer();
			lazyKey.append(font.getFontName().toLowerCase());
			lazyKey.append(hashCode);
			lazyKey.append(split);
			lazyKey.append(text);
		} else {
			lazyKey.delete(0, lazyKey.length());
			lazyKey.append(font.getFontName().toLowerCase());
			lazyKey.append(hashCode);
			lazyKey.append(split);
			lazyKey.append(text);
		}
		return lazyKey.toString();
	}

	/**
	 * 生成一组西方字符缓存键值
	 * 
	 * @param font
	 * @return
	 */
	private static String makeLazyWestKey(LFont font) {
		if (lazyKey == null) {
			lazyKey = new StringBuffer();
			lazyKey.append(font.getFontName().toLowerCase());
			lazyKey.append(font.getStyle());
			lazyKey.append(font.getSize());
		} else {
			lazyKey.delete(0, lazyKey.length());
			lazyKey.append(font.getFontName().toLowerCase());
			lazyKey.append(font.getStyle());
			lazyKey.append(font.getSize());
		}
		return lazyKey.toString();
	}

	/**
	 * 清空西文字体缓存
	 */
	public static void clearEnglishLazy() {
		synchronized (lazyEnglish) {
			for (LSTRFont str : lazyEnglish.values()) {
				if (str != null) {
					str.dispose();
					str = null;
				}
			}
		}
	}

	public static LSTRFont getGLFont(LFont f) {
		if (lazyEnglish.size() > LSystem.DEFAULT_MAX_CACHE_SIZE) {
			clearEnglishLazy();
		}
		String key = makeLazyWestKey(f);
		LSTRFont font = lazyEnglish.get(key);
		if (font == null) {
			font = new LSTRFont(f, true);
			lazyEnglish.put(key, font);
		}
		return font;
	}

	public final static void dispose() {
		clearEnglishLazy();
		clearStringLazy();
	}

}
